src/Analyzers/CSharp/Analyzers/NamespaceFileSync/NamespaceFileSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.NamespaceFileSync
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class NamespaceFileSyncDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private static readonly LocalizableResourceString s_localizableTitle = new LocalizableResourceString(
+          nameof(CSharpAnalyzersResources.Namespace_named_incorrectly), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        private static readonly LocalizableResourceString s_localizableInsideMessage = new LocalizableResourceString(
+            nameof(CSharpAnalyzersResources.Namespace_must_match_folder_structure), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        public const string RootNamespaceOption = "build_property.RootNamespace";
+        public const string ProjectDirOption = "build_property.ProjectDir";
+
+        public NamespaceFileSyncDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId,
+                   CSharpCodeStyleOptions.PreferNamespaceMatchFolderStructure,
+                   LanguageNames.CSharp,
+                   s_localizableInsideMessage)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNamespaceNode, SyntaxKind.NamespaceDeclaration);
+        }
+
+        private void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace)
+                || rootNamespace is null)
+            {
+                return;
+            }
+
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || projectDir is null or { Length: 0 })
```

Comment:
nit: is there a reason this is used as opposed to `string.IsNullOrEmpty()`?

Reply:
I think when I added this I was reading JaredPar's twitter feed. It's neat, but I'm not sure it's more readable for most people yet.

Reply:
yeah.  this really does not read well.